types/src/bytesrepr.rs

```diff
@@ -1317,6 +1087,58 @@ where
     }
 }
 
+/// Serializes dynamic bytes array.
+pub fn serialize_bytes(bytes: &[u8]) -> Result<Vec<u8>, Error> {
+    let serialized_length = bytes_serialized_length(bytes);
+    let mut vec = try_vec_with_capacity(serialized_length)?;
+    let length_prefix = bytes.len() as u32;
+    let length_prefix_bytes = length_prefix.to_le_bytes();
+    vec.extend(&length_prefix_bytes);
+    vec.extend(bytes);
+    Ok(vec)
+}
+
+/// Estimates serialized length of serialized bytes.
+pub fn bytes_serialized_length(bytes: &[u8]) -> usize {
+    U32_SERIALIZED_LENGTH + bytes.len()
+}
+
+/// Deserializes dynamic bytes array.
+pub fn deserialize_bytes(bytes: &[u8]) -> Result<(Vec<u8>, &[u8]), Error> {
+    let (size, remainder) = u32::from_bytes(bytes)?;
+    let (result, remainder) = safe_split_at(remainder, size as usize)?;
+    Ok((result.to_vec(), remainder))
+}
+
+/// Serializes array.
+pub fn serialize_array(bytes: &[u8]) -> Result<Vec<u8>, Error> {
+    Ok(bytes.to_vec())
+}
+
+/// Safely deserializes array of specified length into a slice.
+///
+/// Once const generics are stabilized we can get rid of this helper function.
+pub fn deserialize_array_slice(bytes: &[u8], n: usize) -> Result<(&[u8], &[u8]), Error> {
```

Comment:
Similarly, I think this should be split out into multiple functions so we can have better type safety on the return type.